usr/src/uts/armv8/io/arm_gtmr.c

```diff
@@ -0,0 +1,591 @@
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+/*
+ * Copyright 2017 Hayashi Naoyuki
+ * Copyright 2024 Michael van der Westhuizen
+ */
+
+/*
+ * Resource management for, and implementation of, the Arm Generic Timer Cyclic
+ * Backend.
+ *
+ * This driver and library attaches very early during boot via cbe_init, which
+ * is only concerned with ensuring that this driver is attached. All further
+ * cyclic backend logic is contained in this file and registered by the attach
+ * routine.
+ *
+ * Having the cyclic backend be a proper DDI driver means that implementation
+ * specific interrupt resource management can take effect without leaking
+ * firmware-specific logic into the driver.
+ */
+
+#include <sys/types.h>
+#include <sys/ccompile.h>
+#include <sys/cmn_err.h>
+#include <sys/sunddi.h>
+#include <sys/sunndi.h>
+#include <sys/cyclic.h>
+#include <sys/cyclic_impl.h>
+#include <sys/arch_timer.h>
+#include <sys/smp_impldefs.h>
+#include <sys/clock.h>
+#include <sys/archsystm.h>
+#include <sys/machsystm.h>
+#include <sys/irq.h>
+#include <sys/xc_levels.h>
+
+#define	EXPECTED_NINTRS		4
+#define	ARM_GTMR_INUM_PHYS_NS	1
+#define	ARM_GTMR_INUM_VIRTUAL	2
+
+static ddi_softint_hdl_impl_t cbe_low_hdl =
+	{0, 0, NULL, NULL, 0, NULL, NULL, NULL};
+static ddi_softint_hdl_impl_t cbe_clock_hdl =
+	{0, 0, NULL, NULL, 0, NULL, NULL, NULL};
+
+static hrtime_t cbe_timer_resolution;
+
+static int cbe_ticks = 0;	/* XXXARM: why do we need this? */
+
+/*
+ * cbe_xcall_lock is used to protect the xcall globals since the cyclic
+ * reprogramming API does not use cpu_lock.
+ *
+ * XXXARM: This should not really be necessary - we need to look into how
+ * xcalls are implemented and make them more like other platforms.
+ */
+static kmutex_t cbe_xcall_lock;
+static cyc_func_t volatile cbe_xcall_func;
+static cpu_t *volatile cbe_xcall_cpu;
+static void *cbe_xcall_farg;
+static cpuset_t cbe_enabled;
+
+/*
+ * Cyclic Backend implementation functions
+ *
+ * See the extensive docblock in common/sys/cyclic_impl.h.
+ */
+
+static cyb_arg_t
+cbe_configure(cpu_t *cpu)
+{
+	return (cpu);
+}
+
+static void
+cbe_unconfigure(cyb_arg_t arg __maybe_unused)
+{
+	ASSERT(!CPU_IN_SET(cbe_enabled, ((cpu_t *)arg)->cpu_id));
+}
+
+static void
+cbe_enable(cyb_arg_t arg)
+{
+	processorid_t me = ((cpu_t *)arg)->cpu_id;
+
+	ASSERT((me == 0) || !CPU_IN_SET(cbe_enabled, me));
+	CPUSET_ADD(cbe_enabled, me);
+	arch_timer_set_cval(CNT_CVAL_MAX);
+	arch_timer_unmask_irq();
+	arch_timer_enable();
+}
+
+static void
+cbe_disable(cyb_arg_t arg)
+{
+	processorid_t me = ((cpu_t *)arg)->cpu_id;
+
+	ASSERT(CPU_IN_SET(cbe_enabled, me));
+	arch_timer_disable();
+	arch_timer_mask_irq();
+	CPUSET_DEL(cbe_enabled, me);
+}
+
+/*
+ * XXXARM: Per the CBE block comments we could get a time in the recent past,
+ * which we need to arrange to run ASAP.
+ */
```

Comment:
If this were to happen right now, would we drop the event?  That seems bad, but I think also not actually different from before

Reply:
I need to dig into this to be sure, but using `oneshot_timer_reprogram` (from `i86pc/io/pcplusmp/apic_timer.c`) as a reference, I think we might need to tweak this. I do see this as us as potentially dropping the event at the moment.

I didn't want to hold up this change though, hence the new `XXXARM`.